diff --git a/lib/boot/shims/src/string.js b/lib/boot/shims/src/string.js
index d82f75a..f16e55a 100644
--- a/lib/boot/shims/src/string.js
+++ b/lib/boot/shims/src/string.js
@@ -780,11 +780,7 @@ function h$throwJSException(e) {
   // adding the Exception dictionary
   var strVal = e.toString() + '\n' + Array.prototype.join.call(e.stack, '\n');
   var someE = MK_SOMEEXCEPTION(HS_JSEXCEPTION_EXCEPTION,
-#ifdef GHCJS_PROF
-                               MK_JSEXCEPTION(MK_JSVAL(e), h$toHsString(strVal), h$CCS_SYSTEM))
-#else
                                MK_JSEXCEPTION(MK_JSVAL(e), h$toHsString(strVal))
-#endif
                               );
   return h$throw(someE, true);
 }
